src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -72,195 +72,248 @@ public static FieldType GetFieldDefinition(ISchema schema, IObjectGraphType pare
+                    var valueOutputs = new List<object>(values is ICollection collection ? collection.Count : 0);
+                    int index = 0;
+                    foreach (var val in values)
+                    {
+                        // parse/validate values as required by graph type
+                        valueOutputs.Add(ParseValue(listGraphType.ResolvedType, variableName + "[" + index++ + "]", val));
+                    }
+                    return valueOutputs;
                 }
                 else
                 {
-                    AssertValidVariableValue(schema, listItemType, input, variableName, hasDefaultValue);
+                    // RULE: If the value passed as an input to a list type is not a list and not the null value,
+                    // then the result of input coercion is a list of size one, where the single item value is the
+                    // result of input coercion for the list’s item type on the provided value (note this may apply
+                    // recursively for nested lists).
+                    var result = ParseValue(listGraphType.ResolvedType, variableName, value);
+                    return new List<object>(1) { result };
                 }
-                return;
             }
 
-            if (type is IObjectGraphType || type is IInputObjectGraphType)
+            static object ParseValueObject(IInputObjectGraphType graphType, string variableName, object value)
             {
-                var complexType = (IComplexGraphType)type;
+                if (value == null)
+                    return null;
 
-                if (!(input is Dictionary<string, object> dict))
+                if (!(value is IDictionary<string, object> dic))
                 {
-                    throw new InvalidVariableError(variableName,
-                        $"Unable to parse input as a '{type.Name}' type. Did you provide a List or Scalar value accidentally?");
-                }
+                    // RULE: The value for an input object should be an input object literal
+                    // or an unordered map supplied by a variable, otherwise a query error
+                    // must be thrown.
 
-                // ensure every provided field is defined
-                IList<string> unknownFields = null;
+                    throw new InvalidVariableError(variableName, $"Unable to parse input as a '{graphType.Name}' type. Did you provide a List or Scalar value accidentally?");
+                }
 
-                if (type is IInputObjectGraphType)
+                var newDictionary = new Dictionary<string, object>(dic.Count);
+                foreach (var field in graphType.Fields)
                 {
-                    unknownFields = dict.Keys
-                        .Except(complexType.Fields.Select(f => f.Name))
-                        .ToList();
+                    var childFieldVariableName = variableName + "." + field.Name;
```

Comment:
The same concern here about string concatenation.

Reply:
This will be more tricky to implement without allocations. Either way it will go in a separate PR.